velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -363,6 +366,32 @@ void PageReader::prepareDictionary(const PageHeader& pageHeader) {
           // We start from the end to allow in-place expansion.
           values[i] = parquetValues[i];
         }
+      } else if (
+          type_->type()->isTimestamp() && parquetType == thrift::Type::INT64) {
+        VELOX_CHECK(type_->logicalType_.has_value());
+        auto logicalType = type_->logicalType_.value();
+        if (logicalType.__isset.TIMESTAMP) {
+          VELOX_CHECK(
+              logicalType.TIMESTAMP.isAdjustedToUTC,
+              "Only UTC adjusted Timestamp is supported.");
+          auto values = dictionary_.values->asMutable<Timestamp>();
+          auto parquetValues = dictionary_.values->asMutable<char>();
+          int64_t units;
+
+          if (logicalType.TIMESTAMP.unit.__isset.MICROS) {
+            for (auto i = dictionary_.numValues - 1; i >= 0; --i) {
+              memcpy(&units, parquetValues + i * typeSize, typeSize);
+              values[i] = util::fromUTCMicros(units);
+            }
+          } else if (logicalType.TIMESTAMP.unit.__isset.MILLIS) {
+            for (auto i = dictionary_.numValues - 1; i >= 0; --i) {
+              memcpy(&units, parquetValues + i * typeSize, typeSize);
+              values[i] = util::fromUTCMillis(units);
+            }
+          } else {
+            VELOX_NYI("Nano Timestamp unit is not supported.");
```

Comment:
Is it possible for second unit? Maybe make the error message more accurate.

Reply:
Currently on MICROS, MILLIS, NANOS supported in parquet. For that point error message looks fine.
Or we can remove "Nano" so that it becomes more generic?